A4A > Signup: Adding smoke tests

yashwin opened the pull request:
Closes https://github.com/Automattic/automattic-for-agencies-dev/issues/1136

## Proposed Changes

* This PR adds basic smoke tests to the A4A signup page
* This only checks if some of the form fields are rendered correctly and the values can be entered.

## Testing Instructions

- Switch this branch locally
- Follow the setup instructions here: https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/README.md#quick-start.
- Run `yarn workspace wp-e2e-tests test -- specs/a8c-for-agencies/signup__smoke.ts` and verify the tests pass.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
Add smoke test to A4A signup page

cleacos commented:
I added myself as a reviewer to follow up on this PR... (learning mode)

manzoorwanijk commented on test/e2e/specs/a8c-for-agencies/signup__smoke.ts:
```diff
@@ -0,0 +1,101 @@
+	it( 'Navigate to A4A > Signup', async function () {
+		await page.goto( `${ A4A_URL }/signup` );
+
+		// Enter first name
+		const firstName = 'John';
+		await page.fill( 'input[name="firstName"]', firstName );
+
+		// Enter last name
+		const lastName = 'Doe';
+		await page.fill( 'input[name="lastName"]', lastName );
+
+		// Enter the agency name
+		const agencyName = 'Agency Name';
+		await page.fill( 'input[name="agencyName"]', agencyName );
+
+		// Enter the business URL
+		const businessURL = 'https://example.com';
+		await page.fill( 'input[name="agencyUrl"]', businessURL );
+
+		// Select the number of managed sites
+		await page.selectOption( 'select[name="managed_sites"]', { value: '6-20' } );
+
+		// Select the services offered
+		await page.click( 'input[name="services_offered[]"][value="strategy_consulting"]' );
+
+		// Select the products offered
+		await page.click( 'input[name="products_offered[]"][value="WordPress.com"]' );
```
Just a small feedback here.

For E2E tests, it's [recommended](https://playwright.dev/docs/best-practices#use-locators) to prioritize user-facing attributes and explicit contracts instead of DOM attributes that users do not see. That ensures that the workflow doesn't break even if the DOM structure changes. It also ensures that the page has better a11y.

So, instead of using 

```ts
await page.fill( 'input[name="firstName"]', firstName );
```
It's better to use what the user sees:
```ts
await page.getByLabel( 'First name' ).fill( firstName );
```

CC: @Automattic/team-calypso

manzoorwanijk commented on test/e2e/specs/a8c-for-agencies/signup__smoke.ts:
```diff
@@ -0,0 +1,101 @@
+		);
+
+		// Verify the form values
+		const firstNameValue = await page.inputValue( 'input[name="firstName"]' );
+		expect( firstNameValue ).toBe( firstName );
+
+		const lastNameValue = await page.inputValue( 'input[name="lastName"]' );
+		expect( lastNameValue ).toBe( lastName );
+
+		const agencyNameValue = await page.inputValue( 'input[name="agencyName"]' );
+		expect( agencyNameValue ).toBe( agencyName );
+
+		const businessURLValue = await page.inputValue( 'input[name="agencyUrl"]' );
+		expect( businessURLValue ).toBe( businessURL );
+
+		const managedSitesValue = await page.inputValue( 'select[name="managed_sites"]' );
+		expect( managedSitesValue ).toBe( '6-20' );
+
+		const servicesOfferedValue = await page.inputValue(
+			'input[name="services_offered[]"][value="strategy_consulting"]'
+		);
+		expect( servicesOfferedValue ).toBe( 'strategy_consulting' );
+
+		const productsOfferedValue = await page.inputValue(
+			'input[name="products_offered[]"][value="WordPress.com"]'
+		);
+		expect( productsOfferedValue ).toBe( 'WordPress.com' );
+
+		const countryValue = await page.inputValue(
+			'input[id="components-form-token-input-combobox-control-0"]'
+		);
+		expect( countryValue ).toBe( 'United States' );
```
Also, for assertions, the best practice is to [use web first assertions and avoid manual assertions](https://playwright.dev/docs/best-practices#use-web-first-assertions)

So, instead of 
```ts
const firstNameValue = await page.inputValue( 'input[name="firstName"]' );
expect( firstNameValue ).toBe( firstName );
```
we should do this
```ts
await expect( page.getByLabel( 'First name' ) ).toHaveValue( firstName );
```

CC: @Automattic/team-calypso

manzoorwanijk replied:
It seems that Calypso uses Jest as the test runner for E2E tests instead of Playwright runner. So, web-first assertions may not work.

@Automattic/team-calypso I think we should encourage and add support for web-first assertions for E2E tests by doing this

```bash
yarn workspace wp-e2e-tests add @playwright/test
```

and then importing `expect` and `test` like this

```ts
import { expect, test } from '@playwright/test';

test( 'some test', async ( { page } ) => {
	await expect( page.getByLabel( 'test' ) ).toBeVisible();
} );
```

manzoorwanijk replied:
Also, web-first assertions make tests less flakey.

yashwin committed:
address PR comments

yashwin commented:
Thanks for the review and great feedback, @manzoorwanijk!

I have addressed your comments: https://github.com/Automattic/wp-calypso/pull/94747/commits/dc8eb31ea3cf4006455dc10d33a04b826e2ca363

Could you please take another look at this? I have also updated the testing instructions. 

I'll try implementing your suggestions in my next PR: https://github.com/Automattic/wp-calypso/pull/94747#discussion_r1828851889

manzoorwanijk commented:
Thank you @yashwin for addressing the feedback.

Now you can immediately understand the benefit of using user-facing attributes and how it helps you find a11y issues in the page, because you left the Country option out as it doesn't have an appropriate label or `aria-label` for the combobox, which is a must for users that use assistive technology.

<img width="1765" alt="image" src="https://github.com/user-attachments/assets/77bee552-46ee-4b14-9d50-007c9df16149">

My suggestion would be to attach a label to that combobox and cover it with E2E tests.

manzoorwanijk commented on test/e2e/specs/a8c-for-agencies/signup__smoke.ts:
```diff
@@ -0,0 +1,100 @@
+		// Select the user type to site owner
+		await page
+			.getByRole( 'combobox', { name: 'How would you describe yourself?' } )
+			.selectOption( { value: 'site_owner' } );
```
Here, it will be better to use `label` instead of `value`

```suggestion
			.selectOption( { label: `I do not work at an agency. I'm a site owner.` } );
```

yashwin replied:
Thanks! Fixed it here: https://github.com/Automattic/wp-calypso/pull/94747/commits/1de50a133dca7b07d5523aad70e58ef77c8c4389

yashwin committed:
use label

manzoorwanijk reviewed (approved):
Thank you for making the changes. I like having E2E tests.

yashwin commented:
> Now you can immediately see the benefit of using user-facing attributes and how it helps you find a11y issues in the page, because you left the Country option out as it doesn't have an appropriate label or `aria-label` for the combobox, which is a must for users that use assistive technology.
> My suggestion would be to attach a label to that combobox and cover it with E2E tests.

Yes, this requires some changes at the component level. I have created a task to handle this: https://github.com/Automattic/automattic-for-agencies-dev/issues/1424

yashwin committed:
verify form values

tyxla reviewed (approved):
Nice work on these tests @yashwin 👍 🚀
